### PR TITLE
Duration cleanup

### LINKED ISSUE #174: xs:duration cannot parse minutes without months
Currently the following test fails (**FORG0001**) because the parsing code for `xs:duration` is unable to parse a duration which specifies a minutes field but does not also specify a months field.

``` xquery
(:*******************************************************:)
(: Test: K-SeqExprCast-157                               :)
(: Written by: Frans Englich                             :)
(: Date: 2007-11-22T11:31:21+01:00                       :)
(: Purpose: Ensure that when casting xs:duration to xs:string, that empty components are handled properly. :)
(:*******************************************************:)
xs:string(xs:duration("-PT8H23M0S")) eq "-PT8H23M"
```

### LINKED ISSUE #175: xs:duration fails to validate presence of a field
Currently the following tests are failing because the parsing code for `xs:duration` does not ensure that at least one duration component is present.

``` xquery
(:*******************************************************:)
(: Test: K-SeqExprCast-197                               :)
(: Written by: Frans Englich                             :)
(: Date: 2007-11-22T11:31:21+01:00                       :)
(: Purpose: At least one number and its designator must be present in a xs:yearMonthDuration value. :)
(:*******************************************************:)
xs:yearMonthDuration("P")

(:*******************************************************:)
(: Test: K-SeqExprCast-198                               :)
(: Written by: Frans Englich                             :)
(: Date: 2007-11-22T11:31:21+01:00                       :)
(: Purpose: At least one number and its designator must be present in a xs:yearMonthDuration value, a minus sign is insufficient. :)
(:*******************************************************:)
xs:yearMonthDuration("-P")
```

### LINKED ISSUE #176: xs:duration fails to validate presence of time field
Currently the following test is failing because the `xs:duration` parsing code fails to validate that the time specifier is only included if one or more time fields is present.

``` xquery
(:*******************************************************:)
(: Test: K-SeqExprCast-152                               :)
(: Written by: Frans Englich                             :)
(: Date: 2007-11-22T11:31:21+01:00                       :)
(: Purpose: A 'T' separator in a xs:duration must be followed by time components. :)
(:*******************************************************:)
xs:duration("P1Y24MT")
```

### LINKED ISSUE #177: xs:duration fails to validate complete input
Currently the following tests are failing because the `xs:duration` parsing code fails to validate that the input consumed during parsing is the complete input, with no trailing characters unaccounted for.

``` xquery
(:*******************************************************:)
(: Test: K-SeqExprCast-148                               :)
(: Written by: Frans Englich                             :)
(: Date: 2007-11-22T11:31:21+01:00                       :)
(: Purpose: No zone offset is allowed for xs:duration.   :)
(:*******************************************************:)
xs:duration("P1Y2M123DT10H30M99S+08:00")
```

### LINKED ISSUE #178: xs:yearMonthDuration fails to validate complete input
Currently the following tests are failing because the `xs:yearMonthDuration` parsing code fails to validate that the input consumed during parsing is the complete input, with no trailing characters unaccounted for.

``` xquery
(:*******************************************************:)
(: Test: K-SeqExprCast-201                               :)
(: Written by: Frans Englich                             :)
(: Date: 2007-11-22T11:31:21+01:00                       :)
(: Purpose: '-P3' is an invalid lexical representation for xs:yearMonthDuration. :)
(:*******************************************************:)
xs:yearMonthDuration("-P3")
```

### LINKED ISSUE #179: xs:duration eq comparison ignores differences in month and year
Currently the following tests are failing because the `eq` implementation for `xs:duration` fails to consider the month and year values.

``` xquery
(:*******************************************************:)
(: Test: K-DurationEQ-4                                  :)
(: Written by: Frans Englich                             :)
(: Date: 2007-11-22T11:31:21+01:00                       :)
(: Purpose: Simple test of 'ne' for xs:duration.         :)
(:*******************************************************:)
xs:duration("P1999Y01M3DT08H34M12.143S") ne
           xs:duration("P1999Y10M3DT08H34M12.143S")

(:*******************************************************:)
(: Test: K-DurationEQ-23                                 :)
(: Written by: Frans Englich                             :)
(: Date: 2007-11-22T11:31:21+01:00                       :)
(: Purpose: Simple test of 'ne' for xs:dayTimeDuration with xs:duration on the left hand. :)
(:*******************************************************:)
xs:duration("P1M1DT2H") ne
           xs:dayTimeDuration("P1DT2H")

(:*******************************************************:)
(: Test: K-DurationEQ-24                                 :)
(: Written by: Frans Englich                             :)
(: Date: 2007-11-22T11:31:21+01:00                       :)
(: Purpose: Simple test of 'ne' for xs:dayTimeDuration with xs:duration on the left hand. :)
(:*******************************************************:)
xs:dayTimeDuration("P1DT2H")
            ne xs:duration("P1M1DT2H")
```

### LINKED ISSUE #180: xs:duration values are not ordered
Currently the following tests are failing because the `xs:duration` implementation provides the `lt` and `gt` operators. According to the specification, `xs:yearMonthDuration` and `xs:dayTimeDuration` values are totally ordered, but `xs:duration` is not.

``` xquery
(:*******************************************************:)
(: Test: K-DurationEQ-33                                 :)
(: Written by: Frans Englich                             :)
(: Date: 2007-11-22T11:31:21+01:00                       :)
(: Purpose: The 'ge' operator is not available between xs:duration and xs:dayTimeDuration. :)
(:*******************************************************:)
xs:duration("P1999Y10M3DT08H34M12.143S") ge
               xs:dayTimeDuration("P3DT08H34M12.143S")

(:*******************************************************:)
(: Test: K-DurationEQ-34                                 :)
(: Written by: Frans Englich                             :)
(: Date: 2007-11-22T11:31:21+01:00                       :)
(: Purpose: The 'gt' operator is not available between xs:duration and xs:dayTimeDuration. :)
(:*******************************************************:)
xs:duration("P1999Y10M3DT08H34M12.143S") gt
               xs:dayTimeDuration("P3DT08H34M12.143S")

(:*******************************************************:)
(: Test: K-DurationEQ-35                                 :)
(: Written by: Frans Englich                             :)
(: Date: 2007-11-22T11:31:21+01:00                       :)
(: Purpose: The 'lt' operator is not available between xs:duration and xs:dayTimeDuration. :)
(:*******************************************************:)
xs:duration("P1999Y10M3DT08H34M12.143S") lt
               xs:dayTimeDuration("P3DT08H34M12.143S")

(:*******************************************************:)
(: Test: K-DurationEQ-36                                 :)
(: Written by: Frans Englich                             :)
(: Date: 2007-11-22T11:31:21+01:00                       :)
(: Purpose: The 'le' operator is not available between xs:duration and xs:dayTimeDuration. :)
(:*******************************************************:)
xs:duration("P1999Y10M3DT08H34M12.143S") le
               xs:dayTimeDuration("P3DT08H34M12.143S")
```

### LINKED ISSUE #181: xs:duration should treat -0 and 0 the same
The following tests are currently failing because `xs:duration` fails to normalize -0 to 0.

``` xquery
(:*******************************************************:)
(: Test: K-SeqExprCast-207                               :)
(: Written by: Frans Englich                             :)
(: Date: 2007-11-22T11:31:21+01:00                       :)
(: Purpose: The canonical form of the xs:yearMonthDuration value -P0M is P0M. :)
(:*******************************************************:)
xs:string(xs:yearMonthDuration("-P0M")) eq "P0M"

(:*******************************************************:)
(: Test: K-SeqExprCast-211                               :)
(: Written by: Frans Englich                             :)
(: Date: 2007-11-22T11:31:21+01:00                       :)
(: Purpose: Simple test of casting a negative xs:dayTimeDuration to xs:yearMonthDuration. :)
(:*******************************************************:)

            xs:string(xs:yearMonthDuration(xs:dayTimeDuration("-P31DT3H2M10.001S")))
            eq "P0M"
```

### LINKED ISSUE #182: Incorrect xs:duration conversion to xs:string when negative
The current implementation of `xs:duration` conversion to `xs:string` does not include a leading `-` sign when the following conditions are met:
- The duration is negative
- The duration's only non-zero components are the year, month, and/or minute

This is causing the following tests to fail:

``` xquery
(:*******************************************************:)
(: Test: K-DateTimeFunc-6                                :)
(: Written by: Frans Englich                             :)
(: Date: 2007-11-22T11:31:22+01:00                       :)
(: Purpose: Passing different timezones to fn:dateTime() is an error. :)
(:*******************************************************:)
dateTime(xs:date("2004-03-04-00:01"),
                            xs:time("08:05:23+00:01"))

(:*******************************************************:)
(: Test: K-SeqExprCast-168                               :)
(: Written by: Frans Englich                             :)
(: Date: 2007-11-22T11:31:21+01:00                       :)
(: Purpose: Simple test of casting a negative xs:yearMonthDuration to xs:duration. :)
(:*******************************************************:)

            xs:string(xs:duration(xs:yearMonthDuration("-P543Y456M")))
            eq "-P581Y"
```

### PR DESCRIPTION
**Pull request by sharwell**
Fixes #174 
Fixes #175 
Fixes #176 
Fixes #177 
Fixes #178 
Fixes #179 
Fixes #180 
Fixes #181 
Fixes #182

### CODE REVIEW

**Comment by sharwell** [comment]
|  | Tests Run | Failures | Errors | Skipped |
| --- | --- | --- | --- | --- |
| Before | 13423 | 210 | 62 | 0 |
| After | 13423 | 182 | 59 | 0 |

File: bundles/org.eclipse.wst.xml.xpath2.processor/src/org/eclipse/wst/xml/xpath2/processor/internal/types/XSDuration.java [hunk 1]
```diff
@@ -68,36 +80,54 @@
...
-
-		if (_month >= 12) {
-			_year += _month / 12;
-			_month = _month % 12;
+		assert years >= 0;
+		assert months >= 0;
+		assert days >= 0;
+		assert hours >= 0;
+		assert minutes >= 0;
+		assert seconds != null && seconds.compareTo(BigDecimal.ZERO) >= 0;
```
Comment (RackerWilliams) [inline bundles/org.eclipse.wst.xml.xpath2.processor/src/org/eclipse/wst/xml/xpath2/processor/internal/types/XSDuration.java:88]: Are the asserts in the XSDuration constructor going to bite us?  ...or is the assumtion that there is no way for XSDuration to be called without those conditions holding true

Comment (sharwell) [inline bundles/org.eclipse.wst.xml.xpath2.processor/src/org/eclipse/wst/xml/xpath2/processor/internal/types/XSDuration.java:88]: I don't believe it's possible for these to fail with the current code.

**Comment by RackerWilliams** [comment]
Okay looks good 👍

**Pull request merged**